beacon_node/network/src/sync/block_lookups/mod.rs

```diff
@@ -345,72 +325,83 @@ impl<T: BeaconChainTypes> BlockLookups<T> {
+    /// layer with an `UnknownParent` error.
+    pub fn has_pending_parent_request(&self, block_root: Hash256) -> bool {
+        self.parent_lookups
+            .iter()
+            .any(|parent_lookup| parent_lookup.chain_hash() == block_root)
+    }
+
+    /// Process a block or blob response received from a single lookup request.
+    pub fn single_lookup_response<R: RequestState<Current, T>>(
+        &mut self,
+        lookup_id: SingleLookupReqId,
         peer_id: PeerId,
-        block: Option<Arc<SignedBeaconBlock<T::EthSpec>>>,
+        response: Option<R::ResponseType>,
         seen_timestamp: Duration,
-        cx: &mut SyncNetworkContext<T>,
+        cx: &SyncNetworkContext<T>,
     ) {
-        let stream_terminator = block.is_none().into();
-        let log = self.log.clone();
+        let id = lookup_id.id;
+        let response_type = R::response_type();
 
-        let Some((has_pending_parent_request, request_ref)) = self.find_single_lookup_request(id, stream_terminator, ResponseType::Block) else {
+        let Some(lookup) = self.get_single_lookup::<R>(lookup_id) else {
+            if response.is_some() {
+                warn!(
+                    self.log,
+                    "Block returned for single block lookup not present";
+                        "response_type" => ?response_type,
+                );
+            }
             return;
         };
 
-        let should_remove = match request_ref.verify_block(block) {
-            Ok(Some((block_root, block))) => {
-                if let Some(parent_components) = request_ref.unknown_parent_components.as_mut() {
-                    parent_components.add_unknown_parent_block(block.clone());
-                };
+        let expected_block_root = lookup.block_root();
 
-                if !has_pending_parent_request {
-                    let rpc_block = request_ref
-                        .get_downloaded_block()
-                        .unwrap_or(RpcBlock::new_without_blobs(block));
-                    // This is the correct block, send it for processing
-                    match self.send_block_for_processing(
-                        block_root,
-                        rpc_block,
-                        seen_timestamp,
-                        BlockProcessType::SingleBlock { id },
-                        cx,
-                    ) {
-                        Ok(()) => ShouldRemoveLookup::False,
-                        Err(()) => ShouldRemoveLookup::True,
-                    }
-                } else {
-                    ShouldRemoveLookup::False
-                }
+        match self.single_lookup_response_inner::<R>(peer_id, response, seen_timestamp, cx, lookup)
+        {
+            Ok(lookup) => {
+                self.single_block_lookups.insert(id, lookup);
+            }
+            Err(e) => {
+                debug!(self.log,
+                    "Single lookup retry failed";
```

Comment:
qq - Is it always a retry if we hit this block?

Reply:
no, good point!